### PR TITLE
ambient: revamp "ambient-type" label mechanism

### PR DESCRIPTION
**Pull request by howardjohn**
Fixes https://github.com/istio/istio/issues/43254

Currently, `ambient-type` label has two purposes:
* Signal we want a pod to have redirection, via a controller auto-adding this label
* opt-out of redirection

This was because the initial iteration of the redirection check *at runtime* for this label. The latest CNI, however, has quite different logic; we don't actually need any label at all and can determine redirection by whatever logic we want (ie meshconfig + namespace label).

Additionally, because the logic for "should redirection be enabled" was `ambient-type` label AND <a bunch of other stuff>, things relying on this info (such as XDS controller) need to know all this complex logic.

---

This PR removes the `ambient-type` label and introduces a new annotation, `ambient.istio.io/redirection={enabled,disabled}`. This indicates if redirection is enabled

Most importantly, this annotation is added *by the CNI* when redirection is enabled or disabled. This allows the annotation to act as the source of truth for whether redirection is enabled. Rather than being a signal into whether it _should be_ enabled, the annotation now indicates it *is* enabled. This also removes the need for the "auto label" mechanism

### CODE REVIEW

**Comment by howardjohn** [comment]
> Does this change anything regarding the `istio.io/dataplane-mode=ambient` label?

No, that label configures the namespace. This is about the pod. Users do not need to set this unless they want to opt-out on a per-pod basis. The primary usage is for the CNI to indicate to _other_ components if a pod is in mesh or not

**Comment by linsun** [comment]
Or is this to provide a way for users to annotate the pod directly without namespace being labelled for ambient?

**Comment by linsun** [comment]
> Or is this to provide a (or another?) way for users to annotate the pod directly without namespace being labelled for ambient?

Ok I think your reply answers this :) Basically it used to be the ambient-type label at pod level, but you are proposing to change it to be the pod annotation (`ambient.istio.io/type`).

**Comment by howardjohn** [comment]
The change from label to annotation is mostly for cost and is not the main goal ofthis. The very important distinction is this is set by the CNI as the **source of truth** for whether redirection is set, rather than as a signal of intent. It is like status (actual state) instead of spec (desired state), except we cannot actually write to pod status.

note this is done by many other CNIs

File: pilot/pkg/ambient/ambient.go [hunk 1]
```diff
@@ -17,13 +17,16 @@ package ambient
...
-
-	TypeWorkload NodeType = "workload"
-	TypeNone     NodeType = "none"
+	// AnnotationType determines the type of workload this is.
+	AnnotationType = "ambient.istio.io/type"
+	// TypeMesh indicates this workload is in the mesh and has redirection to ztunnel.
+	// This is added by the CNI when redirection is configured, not by the user.
+	TypeMesh NodeType = "mesh"
+	// TypeDisabled indicates this workload does NOT have redirection enabled. Workloads without any
+	// value should be assumed to be "disabled" as well, but explicitly setting this acts as an opt-out.
```
Comment (bleggett) [inline pilot/pkg/ambient/ambient.go:26]:
```suggestion
	// value should be assumed to be "disabled" as well, but explicitly setting this acts as an opt-out.
```

Is this suggesting that an end-user might set this to `disabled`, or that there is a scenario where the CNI might (versus just not labelling at all)

Comment (howardjohn) [reply pilot/pkg/ambient/ambient.go:26]: CNI will never set it to "disabled", would just not set it. "disabled" implies it was explicitly disabled

File: pilot/pkg/ambient/ambient.go [hunk 2]
```diff
@@ -17,13 +17,16 @@ package ambient
...
+	TypeDisabled NodeType = "disabled"
+	// TypeZtunnel indicates this workload is a ztunnel.
+	TypeZtunnel NodeType = "ztunnel"
+	// TypeWaypoint indicates this workload is a waypoint.
 	TypeWaypoint NodeType = "waypoint"
```
Comment (bleggett) [inline pilot/pkg/ambient/ambient.go:31]:
```suggestion
	TypeZtunnel NodeType = "ztunnel"
	// TypeWaypoint indicates this workload is a waypoint.
	TypeWaypoint NodeType = "waypoint"
```

If this label is purely informational now, and doesn't "do" anything now, aren't states that aren't `mesh` or `<unset>` largely redundant?

i.e. presumably CNI is using $some-other-annotation to decide whether something is a waypoint or ztunnel, and not this - so what's the point of this?

Comment (howardjohn) [reply pilot/pkg/ambient/ambient.go:31]: We could put `ambient.istio.io/type: disabled` on waypoint and ztunnel, but it seems a bit confusing to say they are "disabled". Maybe if we rename the annotation to `ambient.istio.io/redirection: {enabled,disabled}`?

Comment (bleggett) [reply pilot/pkg/ambient/ambient.go:31]: Yeah - I kinda like that better. I agree making this an informational flag that reflects a decision the CNI has already made is a good improvement, and if that's the case really a two-state informational flag.

Comment (howardjohn) [reply pilot/pkg/ambient/ambient.go:31]: k will update

File: cni/pkg/ambient/net.go [hunk 3]
```diff
@@ -144,6 +191,10 @@ func DelPodFromMesh(pod *corev1.Pod) {
 			log.Warnf("Failed to delete route (%s) for pod %s: %v", rte, pod.Name, err)
 		}
 	}
+
+	if err := annotateUnenrollPod(client, pod); err != nil {
+		log.Errorf("failed to annotate pod enrollment: %v", err)
```
Comment (linsun) [inline cni/pkg/ambient/net.go:196]:
```suggestion
		log.Errorf("failed to annotate pod unenrollment: %v", err)
```

File: pilot/pkg/ambient/ambient.go [hunk 4]
```diff
@@ -17,13 +17,16 @@ package ambient
...
-	// TODO this could be an annotation – eventually move it into api repo
-	LabelType = "ambient-type"
-
-	TypeWorkload NodeType = "workload"
-	TypeNone     NodeType = "none"
+	// AnnotationType determines the type of workload this is.
+	AnnotationType = "ambient.istio.io/type"
+	// TypeMesh indicates this workload is in the mesh and has redirection to ztunnel.
+	// This is added by the CNI when redirection is configured, not by the user.
+	TypeMesh NodeType = "mesh"
```
Comment (linsun) [inline pilot/pkg/ambient/ambient.go:24]: mesh could also indicate sidecar too, how about `sidecarless` as the the workload has redirection to ztunnel?

File: pilot/pkg/ambient/ambient.go (hunk 1 above)
Comment (linsun) [reply pilot/pkg/ambient/ambient.go:26]: Seems it should be: `Workloads without any value should NOT be assumed to be "disabled" as well,`

File: pilot/pkg/ambient/ambient.go (hunk 2 above)
Comment (linsun) [reply pilot/pkg/ambient/ambient.go:31]: That could work if the only purpose is for redirection.  I had thought there is other purpose.

Comment (howardjohn) [reply pilot/pkg/ambient/ambient.go:31]: It is only to signal if redirection is enabled. We need a way to signal to things if there is a ztunnel so we know whether to use HBONE basically

File: pilot/pkg/ambient/ambient.go (hunk 4 above)
Comment (bleggett) [reply pilot/pkg/ambient/ambient.go:24]: It's already prefixed with `ambient.istio.io` which should be enough to indicate the scope here is only ambient.

File: pilot/pkg/ambient/ambient.go (hunk 2 above)
Comment (linsun) [reply pilot/pkg/ambient/ambient.go:31]: In fact, we had people don't understand if ingress gateway or waypoint are captured by ztunnel so this would bring some clarity too!

**Review by linsun — COMMENTED** [review]
LGTM, other than a few minor naming changes.

Prefer to have @dhawton to check it out tmr when he is back.

File: pilot/pkg/ambient/ambient.go (hunk 4 above)
Comment (linsun) [reply pilot/pkg/ambient/ambient.go:24]: yeah, except we are having discussion in ambient-dev that ambient also includes sidecar as well.

Comment (howardjohn) [reply pilot/pkg/ambient/ambient.go:24]: We can easily bikeshed the naming later, I hope we will not block development on naming. Its trivial to find+replace it before we ship if we decide to.

Comment (linsun) [reply pilot/pkg/ambient/ambient.go:24]: That is fair - i agree this is a bigger discussion than this PR itself

Comment (bleggett) [reply pilot/pkg/ambient/ambient.go:24]: Something close to what John already suggested: `ambient.istio.io/redirection: {node,sidecar}` might do.

**Comment by howardjohn** [comment]
I've implemented the  on vs off semantics

File: pkg/config/constants/constants.go [hunk 5]
```diff
@@ -159,4 +160,12 @@ const (
 
 	// DataplaneMode namespace label for determining ambient mesh behavior
 	DataplaneMode = "istio.io/dataplane-mode"
+
+	// AmbientRedirection specifies whether a pod has ambient redirection (to ztunnel) configured.
+	AmbientRedirection = "ambient.istio.io/redirection"
+	// AmbientRedirectionEnabled indicates redirection is configured. This is set by the CNI when it
+	// actually sets up redirection, rather than by the user.
+	AmbientRedirectionEnabled = "enabled"
```
Comment (hzxuzhonghu) [inline pkg/config/constants/constants.go:168]: IIUC, from the comment the annotation is set by CNI? If so, we increased calling k8s apiserver in each node. Is there any workaround not calling k8s

File: pilot/pkg/ambient/ambientpod/podutil.go [hunk 6]
```diff
@@ -51,10 +51,9 @@ func ShouldPodBeInIpset(namespace *corev1.Namespace, pod *corev1.Pod, meshMode s
 	return false
 }
 
-// @TODO Interim function for waypoint proxy, to be replaced after design meeting
 func PodHasOptOut(pod *corev1.Pod) bool {
-	if val, ok := pod.Labels["ambient-type"]; ok {
-		return val == "waypoint" || val == "none"
+	if val, ok := pod.Annotations[constants.AmbientRedirection]; ok {
+		return val == constants.AmbientRedirectionDisabled
```
Comment (hzxuzhonghu) [inline pilot/pkg/ambient/ambientpod/podutil.go:56]: So this is excluding pod that is explicitly disabled by user

File: pilot/pkg/serviceregistry/util/label/label.go [hunk 7]
```diff
@@ -33,7 +34,7 @@ const (
 )
 
 // AugmentLabels adds additional labels to the those provided.
-func AugmentLabels(in labels.Instance, clusterID cluster.ID, locality, k8sNode string, networkID network.ID) labels.Instance {
+func AugmentLabels(in labels.Instance, annotations labels.Instance, clusterID cluster.ID, locality, k8sNode string, networkID network.ID) labels.Instance {
```
Comment (hzxuzhonghu) [inline pilot/pkg/serviceregistry/util/label/label.go:37]: This looks a little hacky by putting annotions into labels

**Review by hzxuzhonghu — COMMENTED** [review]
The marjor concern is that CNI operation updating pod

**Comment by hzxuzhonghu** [comment]
>The change from label to annotation is mostly for cost and is not the main goal ofthis

how is that different from labels?

File: pkg/config/constants/constants.go (hunk 5 above)
Comment (howardjohn) [reply pkg/config/constants/constants.go:168]: It's 1 Pod update call for each pod creation. A new pod in k8s already causes >10 API calls, probably much more. Adding annotations in CNI is done by calico, cilium, etc as well

File: pilot/pkg/ambient/ambientpod/podutil.go (hunk 6 above)
Comment (howardjohn) [reply pilot/pkg/ambient/ambientpod/podutil.go:56]: Yes

File: pilot/pkg/serviceregistry/util/label/label.go (hunk 7 above)
Comment (howardjohn) [reply pilot/pkg/serviceregistry/util/label/label.go:37]: Endpoint doesn't have idea of "Annotations" so this seems much simpler than adding a whole new field to Endpoint. Semantically it's the same,

**Comment by howardjohn** [comment]
> > The change from label to annotation is mostly for cost and is not the main goal ofthis
> 
> how is that different from labels?

Labels have a cost in etcd, annotations are intended for these types of purposes

**Comment by birkland** [comment]
I find that issue and associated comments confusing.  So this part seemed clear:

> This PR removes the ambient-type label and introduces a new annotation, `ambient.istio.io/redirection={enabled,disabled}`. This indicates if redirection is enabled

> Most importantly, **this annotation is added by the CNI when redirection is enabled or disabled**. This allows the annotation to act as the source of truth for whether redirection is enabled. **Rather than being a signal into whether it should be enabled, the annotation now indicates it is enabled**. This also removes the need for the “auto label” mechanism

fair enough, but I'm having a difficult time reconciling with:

> we don’t actually need any label at all and can determine redirection by whatever logic we want (ie meshconfig + namespace label).

and

> Does this change anything regarding the `istio.io/dataplane-mode=ambient`?
No, that label configures the namespace. This is about the pod. Users do not need to set this unless they want to opt-out on a per-pod basis. The primary usage is for the CNI to indicate to other components if a pod is in mesh or not

So with the adoption of this proposal, how _does_ a user indicate intent to have a specific pod in the ambient dataplane or not?  (or, for that matter, in the ambient dataplane, in the the sidecar dataplane, or not in the mesh at all)?  Is the label `istio.io/dataplane-mode=ambient` applicable to pods as well as namespaces?  

In other words, 
| User's intent  | Indicated by |
| ------------- | ------------- |
| Include pods from a namespace in ambient mesh |  label  `istio.io/dataplane-mode=ambient` on namespace |
| Exclude  pod from ambient mesh | ??? |
| Include pod in ambient mesh, in namespace not labelled as `istio.io/dataplane-mode=ambient` | ???  |

| What Istio did  | Indicated by |
| ------------- | ------------- |
| Pod in ambient data plane or not | annotation `ambient.istio.io/redirection={enabled,disabled}` |

**Comment by howardjohn** [comment]
@birkland 

With this PR:
| User's intent  | Indicated by |
| ------------- | ------------- |
| Include pods from a namespace in ambient mesh |  label  `istio.io/dataplane-mode=ambient` on namespace |
| Exclude  pod from ambient mesh | annotate pod `ambient.istio.io/redirection=disabled` |
| Include pod in ambient mesh, in namespace not labelled as `istio.io/dataplane-mode=ambient` | Not possible |

| What Istio did  | Indicated by |
| ------------- | ------------- |
| Pod in ambient data plane ~or not~ | annotation `ambient.istio.io/redirection=enabled` |

---

Today:
With this PR:
| User's intent  | Indicated by |
| ------------- | ------------- |
| Include pods from a namespace in ambient mesh |  label  `istio.io/dataplane-mode=ambient` on namespace |
| Exclude  pod from ambient mesh | label pod `ambient-type=none` |
| Include pod in ambient mesh, in namespace not labelled as `istio.io/dataplane-mode=ambient` | Not possible |

| What Istio did  | Indicated by |
| ------------- | ------------- |
| Pod in ambient data plane or not | No action |


---

Note there are only two changes:
* User intent changes from `ambient-type` (which does not meet k8s recommendations) to `ambient.istio.io/redirection`
* CNI exposes the actual state

Other problems with the table (if there are deemed to be any) can be done in a followup.

The primary intent here is that to understand "this pod has redirection", we don't need to replicate the full complex logic the CNI does.

**Review by dhawton — APPROVED** [review]
LGTM

**Comment by birkland** [comment]
@howardjohn 

Thanks for clarifying the labels.  I think that adds a little nuance to the PR description:
> This PR removes the ambient-type label and introduces a new annotation, ambient.istio.io/redirection={enabled,disabled}. This indicates if redirection is enabled

> Most importantly, this annotation is added by the CNI when redirection is enabled or disabled. This allows the annotation to act as the source of truth for whether redirection is enabled. Rather than being a signal into whether it should be enabled, the annotation now indicates it is enabled. This also removes the need for the "auto label" mechanism

In fact, this annotation serves a _dual_ purpose, expressed in your comment:
> * User intent changes from ambient-type (which does not meet k8s recommendations) to ambient.istio.io/redirection
> * CNI exposes the actual state

So now I'm just trying to sort out if there are any repercussions from this dual purpose - would it ever be ambiguous as to who authored the annotation on a given pod?   i.e. if I observe on a pod `ambient.istio.io/redirection=disabled`, how do I know it's annotation from a user expressing intention to opt out, vs the CNI asserting that this pod was excluded from ambient for some reason (could it conceivably do that at some point down the road?)

**Comment by howardjohn** [comment]
The CNI doesn't actually write `disabled`, only `enabled`, so we can avoid modifying workloads that are not our business to touch.

I agree there is some ambiguity though. There was some discussion on the enablement API in general in https://github.com/istio/api/pull/2674 so I was planning to bring it up tomorrow
